File: ts/src/stack/stack.ts
interface StackListNode_Intf {
    value: any;
    next: any
}

interface StackList_Intf {
    append: Function;
    removeLast: Function;
    printElems: Function;
    checkHead: Function;
    getHead: Function;
    getSize: Function;
}

class StackListNode {
    public value: any
    public next: any

    constructor(value: any) {
        this.value = value
        this.next = null
    }
}

class StackList {
    head: StackListNode_Intf
    size: number

    constructor() {
        this.head = null
        this.size = 0
    }

    append(value: any): any {
        var listNode = new StackListNode(value)

        if (this.head) {
            var current = this.head
            while (current.next) {
                current = current.next
            }
            current.next = listNode
        } else {
            this.head = listNode
        }
        this.size += 1
        return listNode
    }

    removeLast(): any {
        if (this.head) {
            var current = this.head, previous = null

            if (this.size === 1) {
                this.head = null
                return current
            } else {

                while (current.next) {
                    previous = current
                    current = current.next
                }

                previous.next = null
                return current.value
            }
        } else return null
    }

    printElems() {
        var result = "______________", current = this.head
        while (current) {
            result += "\n" + current.value
            current = current.next
        }
        console.log(result)
        return result
    }

    checkHead() {
        return this.head === null
    }

    getHead() {
        return this.head.value
    }

    getSize() {
        return this.size
    }
}

class Stack {
    items: StackList_Intf
    constructor() {
        this.items = new StackList()
    }

    push(value: any) {
        return this.items.append(value)
    }

    pop() {
        return this.items.removeLast()
    }

    size() {
        return this.items.getSize()
    }

    isEmpty() {
        return this.items.checkHead()
    }

    peek() {
        return this.items.getHead()
    }

    printStack() {
        return this.items.printElems()
    }
}

export default Stack

File: py/stack/stack.py
class StackListNode:
    def __init__(self, value):
        self.value = value
        self.next = None
    
    def __str__(self):
        return 'Item {}'.format(self.value)

class StackList:
    def __init__(self):
        self.head = None
        self.size = 0

    def __str__(self):
        current , result = self.head, ""
        while current:
            result += str(current.value) + "\n"
            current = current.next
        return result
    
    def append(self, value):
        list_node = StackListNode(value)
        if self.head:
            current = self.head
            while current.next is not None: 
                current = current.next
            current.next = list_node
        else:
            self.head = list_node
        self.size += 1
        return True
    
    def remove_last_item(self):
        if self.head:
            current, previous = self.head, None
            while current.next is not None: 
                previous = current
                current = current.next
            previous.next = None
            self.size -= 1
            return current.value
        else:
            return False
    
    def is_empty(self):
        return self.size == 0
    
    def get_head(self):
        if self.head:
            return self.head.value
        else: return None

    def get_size(self):
        return self.size

    def print_elems(self):
        return self

class Stack:
    def __init__(self):
        self.items = StackList()

    def push(self, value):
        return self.items.append(value)

    def pop(self):
        return self.items.remove_last_item()
    
    def isEmpty(self):
        return self.items.is_empty()
    
    def peek(self):
        return self.items.get_head()

    def size(self):
        return self.items.get_size()
    
    def printStack(self):
        return self.items.print_elems()